crates/ra_ide_api/src/completion/completion_context.rs

```diff
@@ -65,14 +67,23 @@ impl<'a> CompletionContext<'a> {
         Some(ctx)
     }
 
-    // The range of the identifier that is being completed.
-    // This is purely advisory and can be used, for example, to highlight this range in the editor.
-    // Clients are expected to ignore this field.
+    // The range of the identifier that is being completed. Completions should starts with
+    // the range's corresponding text if you want to show in completion candidates.
     pub(crate) fn source_range(&self) -> TextRange {
-        match self.leaf.kind() {
-            // workaroud when completion is triggered by trigger characters.
-            DOT | COLONCOLON => TextRange::from_to(self.offset, self.offset),
-            _ => self.leaf.range(),
```

Comment:
We can just skip all punctuations here, and let each completion function to deal with different contexts. Or we can handle source range here when constructing `CompleteContext`. 

 i don't know which is better.

Reply:
Hm, wouldn't just

```
match self.leaf.kind() {
    IDENT => self.leaf.range(),
    _ => TextRange::offset_let(self.offset, 0.into()),
}
```

work here?

Reply:
don't know there is a `IDENT`.